plenum/test/node_request/test_node_got_no_preprepare.py

```diff
@@ -0,0 +1,129 @@
+
+    # Remove connection problems
+    reset_router_accepting(behind_node)
+
+    # Send txns
+    sdk_send_batches_of_random_and_check(
+        looper, txnPoolNodeSet, sdk_pool_handle, sdk_wallet_client, 3, num_of_batches)
+
+    # behind_node stashing new 3pc messages and not ordering and not participating in consensus
+    assert len(behind_node.master_replica.prePreparesPendingPrevPP) == 1
+    assert not nodes_last_ordered_equal(behind_node, master_node)
+
+    # After DELTA_3PC_ASKING batches, behind_node asks for pre-prepare and starting ordering
+    sdk_send_batches_of_random_and_check(
+        looper, txnPoolNodeSet, sdk_pool_handle, sdk_wallet_client, delta, delta)
+
+    looper.run(eventually(nodes_last_ordered_equal, behind_node, master_node))
+
+
+def test_2_node_got_no_preprepare(looper,
+                                  txnPoolNodeSet,
+                                  sdk_pool_handle,
+                                  sdk_wallet_client,
+                                  tconf):
+    master_node = txnPoolNodeSet[0]
+    behind_nodes = txnPoolNodeSet[-2:]
+    last_ordered = master_node.master_last_ordered_3PC[1]
+    delta = tconf.DELTA_3PC_ASKING
+    num_of_batches = 1
+
+    # Nodes order batches
+    sdk_send_batches_of_random_and_check(
+        looper, txnPoolNodeSet, sdk_pool_handle, sdk_wallet_client, 3, num_of_batches)
+    assert nodes_last_ordered_equal(*behind_nodes, master_node)
+
+    # Emulate connection problems, behind_node doesnt receive pre-prepares
+    router_dont_accept_messages_from(behind_nodes[0], master_node.name)
+
+    # Send some txns and behind_node cant order them while pool is working
+    sdk_send_batches_of_random_and_check(
+        looper, txnPoolNodeSet, sdk_pool_handle, sdk_wallet_client, 3, num_of_batches)
+    assert len(txnPoolNodeSet[1].master_replica.prePrepares) - 1 == \
+           len(behind_nodes[0].master_replica.prePrepares)
+    assert master_node.master_last_ordered_3PC[1] == last_ordered + num_of_batches * 2
+    assert not nodes_last_ordered_equal(behind_nodes[0], master_node)
+
+    # behind_node has requested preprepare and wouldn't request it again until
+    # income preprepare seq_no > last_ordered seq_no + DELTA_3PC_ASKING
+
+    # Remove connection problems
+    reset_router_accepting(behind_nodes[0])
+
+    # Send txns
+    sdk_send_batches_of_random_and_check(
+        looper, txnPoolNodeSet, sdk_pool_handle, sdk_wallet_client, 3, num_of_batches)
+
+    # behind_node stashing new 3pc messages and not ordering and not participating in consensus
+    assert len(behind_nodes[0].master_replica.prePreparesPendingPrevPP) == 1
+    assert not nodes_last_ordered_equal(behind_nodes[0], master_node)
+
+    # Emulate connection problems, behind_node doesnt receive pre-prepares
+    router_dont_accept_messages_from(behind_nodes[1], master_node.name)
+
+    # Send some txns and behind_node cant order them while pool is working
```

Comment:
So, do we see a situation where the pool stopped ordering? If not, then the nodes which lost some 3PC messages may catch-up on stable checkpoint?

Reply:
No, pool will continue write txns in any case and behind nodes will catchup on closest stable checkpoint